src-clojure/me/tonsky/persistent_sorted_set.cljs

```diff
@@ -287,145 +311,273 @@
-                       (.slice new-pointers 0 middle))
-                (Node. (.slice new-keys     middle)
-                       (.slice new-pointers middle)))))))))
-
-  (node-disj [_ cmp key root? left right]
+               (new-node (.slice new-keys     0 middle)
+                         (.slice new-children 0 middle)
+                         (.slice new-addresses 0 middle)
+                         {:address _address})
+               (new-node (.slice new-keys     middle)
+                         (.slice new-children middle)
+                         (.slice new-addresses middle)))))))))
+
+  (node-disj [this cmp key root? left right storage]
+    (ensure-addresses! this (count children))
     (let [idx (lookup-range cmp keys key)]
       (when-not (== -1 idx) ;; short-circuit, key not here
-        (let [child       (arrays/aget pointers idx)
+        (let [child       (node-child this idx storage)
               left-child  (when (>= (dec idx) 0)
-                            (arrays/aget pointers (dec idx)))
-              right-child (when (< (inc idx) (arrays/alength pointers))
-                            (arrays/aget pointers (inc idx)))
-              disjned     (node-disj child cmp key false left-child right-child)]
+                            (node-child this (dec idx) storage))
+              right-child (when (< (inc idx) (arrays/alength children))
+                            (node-child this (inc idx) storage))
+              disjned     (node-disj child cmp key false left-child right-child storage)]
           (when disjned     ;; short-circuit, key not here
             (let [left-idx     (if left-child  (dec idx) idx)
                   right-idx    (if right-child (+ 2 idx) (+ 1 idx))
                   new-keys     (check-n-splice cmp keys     left-idx right-idx (arrays/amap node-lim-key disjned))
-                  new-pointers (splice             pointers left-idx right-idx disjned)]
-              (rotate (Node. new-keys new-pointers) root? left right))))))))
+                  new-children (splice             children left-idx right-idx disjned)
+                  new-addresses (splice            _addresses left-idx right-idx (node-addresses->array disjned))]
+              (when (> right-idx left-idx)
+                (let [cut-addresses (.slice _addresses left-idx right-idx)
+                      removed (set/difference (set (remove nil? cut-addresses)) (set (remove nil? new-addresses)))]
+                  (when (and storage (seq removed))
+                    (protocol/delete storage removed))))
+              (rotate (new-node new-keys new-children new-addresses {:address _address})
+                      root? left right storage))))))))
+
+(declare Leaf)
+(defn- new-leaf
+  [keys & {:keys [address dirty]
+           :or {dirty true}}]
+  (Leaf. keys address dirty))
+
+(deftype Leaf [keys ^:mutable _address ^:mutable _dirty]
+  IStore
+  (store-aux [this storage]
+    (if (or _dirty (nil? _address))
```

Comment:
What is the idea behind `_dirty` flag here? In Clojure version, if node has address, it is persisted. If address is nil, it is not persisted (==dirty?). Do we really need two separate flags for this?